optimizer/evaluator/evaluator.go

```diff
@@ -0,0 +1,850 @@
+		case []byte:
+			f, err := types.StrToFloat(string(x))
+			e.err = err
+			u.SetValue(-f)
+		case mysql.Hex:
+			u.SetValue(-x.ToNumber())
+		case mysql.Bit:
+			u.SetValue(-x.ToNumber())
+		case mysql.Enum:
+			u.SetValue(-x.ToNumber())
+		case mysql.Set:
+			u.SetValue(-x.ToNumber())
+		default:
+			e.err = errors.Errorf("invalid operation")
+		}
+	default:
+		panic("should never happen")
+	}
+
+	return true
+}
+
+func (e *Evaluator) values(v *ast.ValuesExpr) bool {
+	return true
+}
+
+func (e *Evaluator) variable(v *ast.VariableExpr) bool {
+	name := strings.ToLower(v.Name)
+	sessionVars := variable.GetSessionVars(e.ctx)
+	globalVars := variable.GetGlobalVarAccessor(e.ctx)
+	if !v.IsSystem {
+		// user vars
+		if value, ok := sessionVars.Users[name]; ok {
+			v.SetValue(value)
+			return true
+		}
+		// select null user vars is permitted.
+		v.SetValue(nil)
+		return true
+	}
+
+	_, ok := variable.SysVars[name]
+	if !ok {
+		// select null sys vars is not permitted
+		e.err = errors.Errorf("Unknown system variable '%s'", name)
+		return false
+	}
+
+	if !v.IsGlobal {
+		if value, ok := sessionVars.Systems[name]; ok {
+			v.SetValue(value)
+			return true
+		}
+	}
+	value, err := globalVars.GetGlobalSysVar(e.ctx, name)
+	if err != nil {
+		e.err = errors.Trace(err)
+		return false
+	}
+	v.SetValue(value)
+	return true
+}
+
+func (e *Evaluator) funcCall(v *ast.FuncCallExpr) bool {
+	f, ok := builtin.Funcs[v.FnName.L]
+	if !ok {
+		e.err = errors.Errorf("unknown function %s", v.FnName.O)
+		return false
+	}
+	a := make([]interface{}, len(v.Args))
+	for i, arg := range v.Args {
+		a[i] = arg.GetValue()
+	}
+	var argMap map[interface{}]interface{}
+	switch v.FnName.L {
+	case "current_user", "database", "found_rows", "user":
+		argMap = make(map[interface{}]interface{})
+		argMap[builtin.ExprEvalArgCtx] = e.ctx
+	}
+	val, err := f.F(a, argMap)
+	if err != nil {
+		e.err = err
+		return false
+	}
+	v.SetValue(val)
+	return true
+}
+
+func (e *Evaluator) funcExtract(v *ast.FuncExtractExpr) bool {
+	val := v.Date.GetValue()
+	if val == nil {
+		v.SetValue(nil)
+		return true
+	}
+
+	f := types.NewFieldType(mysql.TypeDatetime)
+	f.Decimal = mysql.MaxFsp
+	var err error
+	val, err = types.Convert(val, f)
+	if err != nil {
+		e.err = errors.Trace(err)
+		return false
+	}
+	if val == nil {
+		v.SetValue(nil)
+		return true
+	}
+
+	t, ok := val.(mysql.Time)
+	if !ok {
+		e.err = errors.Errorf("need time type, but got %T", val)
+		return false
+	}
+	n, err1 := mysql.ExtractTimeNum(v.Unit, t)
+	if err1 != nil {
+		e.err = errors.Trace(err1)
+		return false
+	}
+	v.SetValue(n)
+	return true
+}
+
+func (e *Evaluator) funcConvert(f *ast.FuncConvertExpr) bool {
+	value := f.Expr.GetValue()
+
+	// Casting nil to any type returns nil
+	if value == nil {
+		return true
+	}
+	str, ok := value.(string)
+	if !ok {
+		return true
+	}
+	if strings.ToLower(f.Charset) == "ascii" {
+		f.SetValue(value)
+		return true
+	} else if strings.ToLower(f.Charset) == "utf8mb4" {
```

Comment:
I think upper case "ASCII" and "utf8mb4" can not be looked up in the encoding table, so it is set as it is.